### DISCUSSION TITLE
Zoom into a pdf within lambda and having issues.

### DISCUSSION THREAD

**Discussion by meggievh** [pymupdf/PyMuPDF#1598] [Q&A]
Essentially the copy I create is exactly the same as the original copy. I really just want to at least zoom in such that the border margins would go away. 

   file_obj = event["Records"][0]
    bucketname = str(file_obj["s3"]["bucket"]["name"])
    filename = unquote_plus(str(file_obj["s3"]["object"]["key"]))
    
    doc = fitz.open()
    s3 = boto3.resource('s3')
    obj = s3.Object(bucketname, filename)
    fs = obj.get()['Body'].read()
    pdf=fitz.open("pdf", stream=BytesIO(fs))
     # open stream as PDF
    rect = pdf[0].rect
    sizeA4 = fitz.paper_size("A4")
    page = doc.new_page(width = sizeA4[0], height = sizeA4[1])
    page.show_pdf_page(rect, pdf, 0)
  
    new_bytes=doc.write()

    
    
  
    bucketname2='modified'
    s3.Bucket(bucketname2).put_object(Key=filename, Body=new_bytes)

**Comment by JorjMcKie** [comment]
Looks like this is more a question than an issue.
So I will transfer this to the `Discussions` tab.

**Comment by JorjMcKie** [comment]
Ok, now.
I am still confused as to what are you trying to achieve:
You have some PDF page that you don't want to show fully?

**Comment by JorjMcKie** [comment]
If this is the case, you could simply decrease the page's cropbox. E.g. do this:
```python
r = page.rect + (72, 72, -72, -72)
# the above rect omits 1 inch (= 72 points) from all the borders
page.set_cropbox(r)
# done
```

**Comment by meggievh** [comment]
I have a pdf scanned image that is small in the right hand corner of the
page. When I zoom in locally  using os.system(f'pdf-crop-margins {file_pdf}
-o {outputpdf}') it removes margins and the OCR works much better. I want
to do something similar using fitz (as it is easier to import into a
lambda).

Thank you!

**Reply by JorjMcKie** [reply]
Ah, ok. Would my previous post help with this?

**Comment by meggievh** [comment]
I tried the  code and I think I need something maybe that zooms in to get
the similar effect as the other package.
 r = page.rect + (72, 72, -72, -72)

# the above rect omits 1 inch (= 72 points) from all the borders
page.set_cropbox(r)

**Comment by JorjMcKie** [comment]
What you could also do is taking a (high resolution) RGB Pixmap of the page's wanted part and make an OCR from this:
```python
r= page.rect + (....)
pix = page.get_pixmap(dpi=300, clip=r)
pdfbytes = pix.pdfocr_tobytes()
ocrpdf = fitz.open("pdf", pdfbytes)
ocrpage=ocrpdf[0]
# now extract your text
text = ocrpage.get_text()
# or any of the other get_text() variants
```

**Reply by JorjMcKie** [reply]
The dpi parameter is equivalent to zooming.
I am not sure whether you indicated that also need to distort the scan, because the original was not positioned completely flat on the scanning machine.
If so, then you maybe could use the new Pixmap method `warp()` to make a rectangle again from something that looks more like a trapezoid ...

**Comment by JorjMcKie** [comment]
If you need text position coordinates with respect to the original scanned page, you can also compute them fairly easily ... let me know.

**Comment by meggievh** [comment]
Essentially I am a bit stuck on using this within a lambda too. I need the
zoomed image to be a pdf  in order to leverage textract.
I must have deleted the code too that made this save that that doc.write
works

def lambda_handler(event, context):

    textract = boto3.client("textract")
    #if event:
    file_obj = event["Records"][0]
    bucketname = str(file_obj["s3"]["bucket"]["name"])
    filename = unquote_plus(str(file_obj["s3"]["object"]["key"]))

    doc = fitz.open()
    s3 = boto3.resource('s3')
    obj = s3.Object(bucketname, filename)
    fs = obj.get()['Body'].read()
    pdf=fitz.open("pdf", stream=BytesIO(fs))
     # open stream as PDF
    page = pdf[0]
    #sizeA4 = fitz.paper_size("A4")

    r = page.rect + (72, 72, -72, -72)

  #the above rect omits 1 inch (= 72 points) from all the borders
    page.set_cropbox(r)
    #page.show_pdf_page(rect, pdf, 0)



    new_bytes=doc.write()

**Comment by JorjMcKie** [comment]
>  I need the zoomed image to be a pdf  in order to leverage textract.

But my snippet using this pixmap approach, does produce an (intermediate) PDF! With one page that contains the original page image (with cropped border) and an underlying OCRed text layer.

You also could stick with your original PDF page, too. Set the page's cropbox as described and then do this to OCR and extract the text:
```python
page.set_cropbox(r)
tp = page.get_textpage_ocr(dpi=300, full=True)
# the above performs the OCR, then makes a textpage to extract from:
text = page.get_text("text", textpage=tp)
# or other text output formats
blocks = page.get_text("dict", textpage=tp)["blocks"]
# etc.
```
This produces the same result as working with pixmaps.

**Comment by meggievh** [comment]
I am using textract so I need to have just a zoomed in version of the pdf.
I have tried this with other image formats and did not work as well.  I am
able to crop with code, but it is not zoomed in.

**Comment by JorjMcKie** [comment]
I am beginning to suspect, that what you refer to as `textract` is something outside PyMuPDF - correct?
Else I don't understand your problem: if you use the dpi parameter in one of the above ways, together with cropping/clipping the page area, you **_are zooming_**.

**Comment by meggievh** [comment]
My reason for using fitz is to just take in a pdf, zoom in, and then resave
it zoomed in, so I can later use other OCR packages.

When
would I use cover_to_pdf on a pixmaps. That is the step I am missing I
believe
pdfbytes = doc.convert_to_pdf() # this a bytes object

I would prefer to not have to save in s3 then reread in

**Comment by JorjMcKie** [comment]
Ah, this is a completely different thing!
If you have a scanned pdf (i.e. every page is an image), you can make a zoomed version of each page and save to a new PDF with thus modified pages.
```python
outpdf = fitz.open()
for page in inpdf:
    pix = page.get_pixmap(dpi=300, clip=...)  # make a zoomed in version of the relevant page part
    outpage = outpdf.new_page(width=pix.width, height=pix.height)
    outpage.insert_image(outpage.rect, pixmap=pix)
outpdf.close(...,garbage=4,deflate=True)
```
